### PR TITLE
[ML] only persist progress if it has changed

### LINKED ISSUE #61913: ClassificationIT.testTooLowConfiguredMemoryStillStarts fails on CI
The test `ClassificationIT.testTooLowConfiguredMemoryStillStarts` fails regularly on CI on both `master` and `7.x` branches.

**Build scan**:
https://gradle-enterprise.elastic.co/s/lb4svvfnvgrgg

**Repro line**:

**Reproduces locally?**:
No

**Applicable branches**:
`master` and `7.x`

**Failure history**:

[Build stats for last 30 days](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:'2020-07-05T13:38:09.491Z',mode:absolute,to:'2020-09-03T13:38:09.492Z'))&_a=(columns:!(branch,test),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:e58bf320-7efd-11e8-bf69-63c8ef516157,key:test,negate:!f,params:(query:testTooLowConfiguredMemoryStillStarts,type:phrase),type:phrase,value:testTooLowConfiguredMemoryStillStarts),query:(match:(test:(query:testTooLowConfiguredMemoryStillStarts,type:phrase))))),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'class:%22org.elasticsearch.xpack.ml.integration.ClassificationIT%22'),sort:!(time,desc)))

**Failure excerpt**:
It looks like the test leaves transient cluster  metadata but I'm not sure why. The most notable `AssertionError` is

```
org.elasticsearch.xpack.ml.integration.ClassificationIT > testTooLowConfiguredMemoryStillStarts FAILED
    java.lang.AssertionError: 
    Expected: <0L>
         but: was <1L>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.cleanUpAnalytics(MlNativeDataFrameAnalyticsIntegTestCase.java:96)
        at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.cleanUpResources(MlNativeDataFrameAnalyticsIntegTestCase.java:87)
        at org.elasticsearch.xpack.ml.integration.MlNativeIntegTestCase.cleanUp(MlNativeIntegTestCase.java:181)
        at org.elasticsearch.xpack.ml.integration.ClassificationIT.cleanup(ClassificationIT.java:111)

    java.lang.AssertionError: test leaves transient cluster metadata behind
    Expected: an empty collection
         but: <[logger.org.elasticsearch.xpack.ml.dataframe.inference, logger.org.elasticsearch.xpack.core.ml.inference]>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:545)
        at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2054)
```

but skimming through the logs also show some ERROR messages like:

```
» ERROR][o.e.x.m.d.p.AnalyticsProcessManager] [integTest-0] [test_model_memory_limit_lower_than_estimated_memory_usage] Marking task failed; [test_model_memory_limit_lower_than_estimated_memory_usage] analytics process stopped unexpectedly: Input error: memory limit 1MB is too low to perform analysis. You need to give the process at least 4MB, but preferably more.
```

### PR DESCRIPTION
**Pull request by benwtrent**
We already search for the previously stored progress document.

For optimization purposes, and to prevent restoring the same
progress after a failed analytics job is stopped,
this commit does an equality check between the previously stored progress and current progress
If the progress has changed, persistence continues as normal.

closes https://github.com/elastic/elasticsearch/issues/61913

### CODE REVIEW

File: x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlNativeDataFrameAnalyticsIntegTestCase.java [hunk 1]
```diff
@@ -93,7 +97,13 @@ private void cleanUpAnalytics() {
         for (DataFrameAnalyticsConfig config : analytics) {
             try {
                 assertThat(deleteAnalytics(config.getId()).isAcknowledged(), is(true));
-                assertThat(searchStoredProgress(config.getId()).getHits().getTotalHits().value, equalTo(0L));
+                if (searchStoredProgress(config.getId()).getHits().getTotalHits().value > 0) {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlNativeDataFrameAnalyticsIntegTestCase.java:100]: Could you explain how we could have leftover progress here?

Comment (benwtrent) [reply x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlNativeDataFrameAnalyticsIntegTestCase.java:100]: @dimitris-athanasiou the overall change only partially handles the race condition window.

- force close a job
- persist progress
- delete job (deletes previous progress)
- job finally fires onCancelled()
- job sees that there was no previous progress persisted
- persists the new progress

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlNativeDataFrameAnalyticsIntegTestCase.java:100]: OK, I think we can then remove the assertion here entirely. I'll add a test dedicated to deleting a job and checking things like the stored progress doc was cleaned up in a separate PR.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Pull request merged**